frame/paged-list/src/paged_list.rs

```diff
@@ -0,0 +1,530 @@
+	_phantom: PhantomData<(Prefix, Hasher, Value, ValuesPerPage, MaxPages)>,
+}
+
+/// The state of a [`StoragePagedList`].
+///
+/// This struct doubles as [`frame_support::storage::StorageList::Appendix`].
+#[derive(
+	Encode, Decode, CloneNoBound, PartialEqNoBound, EqNoBound, DebugNoBound, DefaultNoBound,
+)]
+// todo ignore scale bounds
+pub struct StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage> {
+	/// The first page that contains a value.
+	///
+	/// Can be >0 when pages were deleted.
+	pub first_page: PageIndex,
+	/// The first value inside `first_page` that contains a value.
+	///
+	/// Can be >0 when values were deleted.
+	pub first_value: ValueIndex,
+
+	/// The last page that could contain data.
+	///
+	/// Iteration starts at this page index.
+	pub last_page: PageIndex,
+	/// The last value inside `last_page` that could contain a value.
+	///
+	/// Iteration starts at this index. If the page does not hold a value at this index, then the
+	/// whole list is empty. The only case where this can happen is when both are `0`.
+	pub last_value: ValueIndex,
+
+	_phantom: PhantomData<(Prefix, Hasher, Value, ValuesPerPage)>,
+}
+
+impl<Prefix, Hasher, Value, ValuesPerPage> frame_support::storage::StorageAppendix<Value>
+	for StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage>
+where
+	Prefix: StorageInstance,
+	Value: FullCodec,
+	Hasher: StorageHasher,
+	ValuesPerPage: Get<u32>,
+{
+	fn append<EncodeLikeValue>(&mut self, item: EncodeLikeValue)
+	where
+		EncodeLikeValue: EncodeLike<Value>,
+	{
+		self.append_one(item);
+	}
+}
+
+impl<Prefix, Hasher, Value, ValuesPerPage>
+	StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage>
+where
+	Prefix: StorageInstance,
+	Value: FullCodec,
+	Hasher: StorageHasher,
+	ValuesPerPage: Get<u32>,
+{
+	pub fn from_storage() -> Option<Self> {
+		let key = Self::key();
+
+		sp_io::storage::get(key.as_ref()).and_then(|raw| Self::decode(&mut &raw[..]).ok())
+	}
+
+	pub fn key() -> Hasher::Output {
+		(StoragePagedListPrefix::<Prefix>::final_prefix(), b"meta").using_encoded(Hasher::hash)
+	}
+
+	pub fn append_one<EncodeLikeValue>(&mut self, item: EncodeLikeValue)
+	where
+		EncodeLikeValue: EncodeLike<Value>,
+	{
+		// Note: we use >= here in case someone decreased it in a runtime upgrade.
+		if self.last_value >= ValuesPerPage::get() {
+			self.last_value = 0;
+			self.last_page.saturating_inc();
+		}
+		let key = page_key::<Prefix, Hasher>(self.last_page);
+		self.last_value.saturating_inc();
+		// Pages are `Vec` and therefore appendable.
+		sp_io::storage::append(key.as_ref(), item.encode());
+		self.store();
+	}
+
+	pub fn store(&self) {
+		let key = Self::key();
+		self.using_encoded(|encoded| sp_io::storage::set(key.as_ref(), encoded));
+	}
+
+	pub fn reset(&mut self) {
+		*self = Default::default();
+		Self::clear();
+	}
+
+	pub fn clear() {
+		let key = Self::key();
+		sp_io::storage::clear(key.as_ref());
+	}
```

Comment:
I renamed the `clear` to `delete`.  

About the `reset`: it is a trick to easily access the type of `StoragePagedListMeta`. Otherwise i need to type out `StoragePagedListMeta<Prefix, Hasher, Value, ValuesPerPage>::delete()` just to access the `delete` function. And typing that out could be error prone…